File: powers/binary_search.py
"""
Module with binary search implementations.
"""
from typing import List


def binary_search(nums: List[int], target: int) -> int:
    l = 0
    r = len(nums) - 1

    while l <= r:
        mid = (l + r) // 2

        if nums[mid] == target:
            return mid
        elif nums[mid] > target:
            r = mid - 1
        else:
            # this part can be changed to find upper and lower bounds
            # by keep binary searching on left/right subarrays
            l = mid + 1

    return -1

File: challenges/leetcode/binary_search.cpp
/*
 * Solution for LC#704: Binary Search
 *
 * https://leetcode.com/problems/binary-search/
 */
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int search(vector<int>& nums, int target) {
        int l = 0;
        int r = nums.size() - 1;

        // reducing the search space
        while (l <= r) {
            int mid = (l + r) / 2;

            if (nums[mid] < target) {
                l = mid + 1;
            } else if (nums[mid] > target) {
                r = mid - 1;
            } else {
                return mid;
            }
        }

        return -1;
    }
};
